integrated_channels/canvas/client.py

```diff
@@ -51,9 +51,123 @@ def __init__(self, enterprise_configuration):
         self.expires_at = None
 
     def create_course_completion(self, user_id, payload):  # pylint: disable=unused-argument
-        pass
+        learner_data = json.loads(payload)
+        self._create_session()
+
+        # Retrieve the Canvas user ID from the user's edx email (it is assumed that the learner's Edx
+        # and Canvas emails will match).
+        get_user_id_from_email_url = '{url_base}/api/v1/accounts/{account_id}/users?search_term={email_address}'.format(
+            url_base=self.enterprise_configuration.canvas_base_url,
+            account_id=self.enterprise_configuration.canvas_account_id,
+            email_address=user_id
+        )
+        rsps = self.session.get(get_user_id_from_email_url)
+        rsps.raise_for_status
+        get_users_by_email_response = rsps.json()
+        try:
+            canvas_user_id = get_users_by_email_response[0]['id']
+        except (KeyError, IndexError):
+            # Trying to figure out how we should handle errors here- should we catch
+            # them in the in the transmitter? Should we return 400 but not raise?
+            # If we have multiple course completions to post, a raised exception here without
+            # anything else will prevent other transmissions to happen.
+            raise CanvasClientError(
+                "No Canvas user ID found associated with email: {}".format(learner_data['userID'])
+            )
+
+        # With the Canvas user ID, retrieve all courses for the user.
+        get_users_courses_url = '{canvas_base_url}/api/v1/users/{canvas_user_id}/courses'.format(
+            canvas_base_url=self.enterprise_configuration.canvas_base_url,
+            canvas_user_id=canvas_user_id
+        )
+        rsps = self.session.get(get_users_courses_url)
+        rsps.raise_for_status()
+        user_courses_response = rsps.json()
+
+        # Find the course who's integration ID matches the learner data course ID
+        # Is there a better identifier here?
+        course_id = None
+        for course in user_courses_response:
+            integration_id = course['integration_id']
+            if '+'.join(integration_id.split(":")[1].split("+")[:2]) == learner_data['courseID']:
+                course_id = course['id']
+                break
+
+        if not course_id:
+            raise CanvasClientError(
+                "Course: {course_id} not found registered in Canvas for Edx learner: {user_id}"
+                "/Canvas learner: {canvas_user_id}.".format(
+                    course_id=learner_data['courseID'],
+                    user_id=learner_data['userID'],
+                    canvas_user_id=canvas_user_id
+                ))
+
+        # Check if the course assignment already exists
+        canvas_assignments_url = '{canvas_base_url}/api/v1/courses/{course_id}/assignments'.format(
```

Comment:
I'm checking integration id of the assignment below this for just this case! If the course already has a bunch of assignments, then it should only deal with one's that have been created here in the client, ie 
```        if not assignment_id:
            assignment_creation_data = {
                'assignment': {
                    'name': 'Final Grade',
                    'submission_types': 'none',
                    'integration_id': integration_id,
                    'published': True,
                    'points_possible': 100
                }
            }```
The one thing I am now realizing is that this "learner data" assignment is labeled as "Final Grade" and reflects the user's grade as such for the course, but acts as a single assignment on Canvas which could potentially create some confusion with other assignments because each will be weighted towards the **Canvas'** final grade. I am not sure how to get around this though.

Reply:
yup let's talk about this in the team! Perhaps Ken also have some thoughts on this